### ISSUE TITLE
`managed.Reconciler` leaks resources with non-deterministic IDs

### ISSUE THREAD

**Issue by dee0** [crossplane/crossplane-runtime#277]
### What happened?
After deleting a composite and waiting, with kubectl wait,  for all of the referenced api objects to be deleted, managed resources still existed in AWS.
This is a problem for us because this wait logic is a gate used to delay deletion of our k8s cluster.  If the managed resources in AWS aren't cleaned up by the time we are done waiting they will most likely end up orphaned.  ( We are seeing this often )

### How can we reproduce it?
Here is the clean up logic we are performing before allowing the cluster to be deleted
Get all CompositeResourceDefinition
For each CompositeResourceDefinition
  If .spec.claimNames.kind
    Get all claims of .spec.claimNames.kind
    For claim
      Get resource refs
      Delete claim
    Wait for all resources to be deleted
  For each composition of .spec.names.kind
    Get resource refs
    Delete composition
    Wait for all resources to be deleted
Get all managed resources ( kubectl get managed )
  For each managed resource
    Delete managed resource

We expect that once we have run through the above steps all managed resources have been deleted from AWS.
However this doesn't seem to be the case.

In the attached zip are example 
xrd.yaml
composition.yaml
claim.yaml
used to create the resources in AWS.

Also in the zip is events-for-vpc-14449-d66-dmi.xlsx which shows the events around one of the leaked VPC.
Events at top are most recent.
Events with white background are from CloudTrails
Events with light blue background are from my team's clean up code.
Events with green background are from Gardener

At line 81 you can see the first event from my team's clean up code.
By line 31, all of the resources related to the VPC have been cleaned up from k8s.   There is nothing left for our clean up code to delete or wait for.
Our last attempt to wait for a resources to be cleaned up is about 9 seconds after the last event for the VPC in CloudTrails.   Given all of the deletion attempts that failed with Client.DependencyViolation I am wondering if Crossplane has given up.   ( And why are all those failures there in the first place? )
Anyhoo once our cleanup code has nothing left to wait for cluster Gardener is free to begin with the cluster deletion.   You can see at line 28, about 45 seconds after my team's clean up code
finished, where it begins taking destructive action.   It starts with  simply going through the api server and deleting all CRD that it can find.
Since Gardener is just blindly deleting things if the AWS resources haven't been cleaned up at this point it is highly likely they are going to be orphaned.
[14449-d66-dmi-leak.zip](https://github.com/crossplane/crossplane/files/6812143/14449-d66-dmi-leak.zip)

### What environment did it happen in?
crossplane:v1.2.2
aws provider provider-aws:v0.18.1
kubernetes 1.19.10

**Comment by negz** [comment]
I'm going to self assign this in the (possibly optimistic) hope that I'll get some time to look into this next week.

**Comment by negz** [comment]
Currently I'm trying to reproduce this issue. If I can, I’m thinking of leveraging the optimistic concurrency control we get on writes to the API server to try to address the issue. For the handful of RouteTable-like resources we have I’m hoping we could change the flow to be:

1. Write an annotation like `crossplane.io/external-name-pending: $timestamp` to the relevant resource.
2. Create the resource in the external API (e.g. AWS).
3. Persist the crossplane.io/external-name annotation like we do today.

This won’t prevent us from leaking external resources with non-deterministic IDs, because that’s pretty much impossible to prevent AFAICT. It’s always possible that we’ll error out while persisting the external-name annotation and need to requeue. However by writing an annotation before we try create we can guarantee that we have the latest version of the (e.g.) RouteTable because the API server will fail the write if we don’t. We can also avoid transparently creating the RouteTable twice and at least warn the user that we potentially leaked one.

**Comment by negz** [comment]
@dee0sap do you have any tips on how I might reproduce this? I'm using the below XR and Composition that seem roughly the same as yours, but I'm yet to find a leaked route table.

https://gist.github.com/negz/e1f2e74f18802d15440214a1a1abc981

**Comment by dee0sap** [comment]
@negz Yeah, your XR and composition seem the same ( at least at a glance )

Maybe it is a question of a scale?   In our case this code is being run by 
- Many devs in Germany and North America.   Often they will run it when around when they come online.   Really not sure of the numbers, but lets say 50+ in DE and 50+ and NA.   They will all be hitting the same AWS account.     
Fwiw what is going on is they are setting up test clusters for themselves.   It takes about 45 minutes for the process to complete and the clusters only live for 24 hours.
- CI/CD pipelines that are running integration tests.    We have probably 50+ micrososervices.   For many of those, the git repos have branch protection checks that run integration tests on open PRs.   And for all of them, once code is merged into master integration tests run which execute the code several times ( lets say 5 ).   This activity happens in a different AWS account than dev activity.    Similar to above, what is happening is clusters are being created and destroyed for the purposes of running integration tests.

That said, at least in the case of the set of logs I sent you last, and that have spent the most time staring at, I saw that the route table had to be requeued at least a few times because the referenced resources ( vpc, natgateway ) weren't available yet.   Not sure that is relevant or how you could force it to happen if it is.

Are you trying to reproduce the situation 'for real' or are you using mocks within a unit test?

**Comment by negz** [comment]
I _believe_ I've reproduced this (for real) using the script above, applying `leaktest.yaml` against an AWS account. I didn't keep count, but I created and destroyed ~1,000 instances of the `LeakTest` over the course of ~10 hours. When I checked in this morning I notice two (of ~1,000) VPCs were not able to be deleted due to extra RouteTables. I need to confirm that these were created by Crossplane, but I expect they were.

**Comment by dee0sap** [comment]
Cool! :)

Btw I in our case the extra route tables 
- are missing all the tags you would expect them to have.  e.g. owner is missing
- are nameless 
- are _not_ main route tables

So I can spot suspects quickly just be looking for nameless, non-main, route tables.    I then check for no tags and check cloud trails to see if the deployment user created it.   If I see that then I know I am looking at a leak.

**Comment by negz** [comment]
That makes sense. I unfortunately don't have CloudTrail enabled so I can't use that to confirm, but these routes are missing tags. Missing tags makes sense because Crossplane actually adds tags after creation in an update pass for this particular resource. When you mention routes being nameless, are you referring to the "Name" tag?

**Comment by dee0sap** [comment]
![image](https://user-images.githubusercontent.com/82829708/129277409-02aa48e6-3a09-4616-865e-1a08e665bfbe.png)

See the screenshot.   Think that is the same as the name tag but I am not sure.

**Comment by negz** [comment]
Hah. I looked all over in the console and somehow missed that column that was right in front of me. I just confirmed the name is indeed just the value of the `Name` tag.

**Comment by negz** [comment]
I'm pretty confident what we're seeing is actually eventual consistency in the AWS API. I do think the crossplane-runtime issue we identified is real, but I haven't actually observed it in practice. I noticed the following log lines while testing a potential fix (https://github.com/crossplane/crossplane-runtime/pull/279):

```
2021-08-14T08:37:33.972Z        DEBUG   provider-aws    Successfully requested creation of external resource    {"controller": "managed/routetable.ec2.aws.crossplane.io", "request": "/test-1-s6hcj", "uid": "ff9f2de0-866e-41d1-956a-6719578f15fb", "version": "221056", "external-name": ""}
2021-08-14T08:37:34.751Z        DEBUG   provider-aws    Successfully requested creation of external resource    {"controller": "managed/routetable.ec2.aws.crossplane.io", "request": "/test-1-s6hcj", "uid": "ff9f2de0-866e-41d1-956a-6719578f15fb", "version": "221492", "external-name": "rtb-0673d041d56da694f"}
```

What's interesting about this is that the value of `"external-name"` is actually set _before_ the `Create` call, and not updated after. So the second log line indicates that we already had our external name annotation set to `rtb-0673d041d56da694f`, but when we called the `RouteTable` `Observe` method it reported that that route table `rtb-0673d041d56da694f` did not exist. I suspected this could be due to the `DescribeRouteTablesRequest` AWS call sometimes lagging a little behind, and it appears that's the case. You can see in the below Terraform PR that they demonstrated the issue and  built in some retries in order to wait for the route table to exist:

https://github.com/hashicorp/terraform-provider-aws/pull/1791#issue-144126468

I suspect the fix here will be to add a reasonable amount of retries with exponential backoff around the `DescribeRouteTablesRequest` call in the `RouteTable` controller.
